### PR TITLE
Let the developers use a different response type for limitation message

### PR DESCRIPTION
**Pull request by HorlogeSkynet**
Hello back,

In order to handle the case specified in #26, I've just allowed a propagation of `type` configuration field to `Response` objects.
It's quicker and surely cleaner as I would have done with a extra headers solution.

Tell me if it's OK for you, or if I should take over a part globally...

Maybe we should change the `message` field instead ?

```php
'message' => [
    'body' => 'Rate limit exceeded',
    'type' => 'text/html'
]
```

Bye

### CODE REVIEW

**Comment by ADmad** [comment]
Perhaps this would be even better:
```php
'message' => [
    'body' => 'Rate limit exceeded',
    'headers' => [
        'Content-type' => 'text/html'
    ]
]
```

This would allow users to set any additional headers they need not just content type.

**Comment by ADmad** [comment]
Also instead of 'message' could use key 'response'.

**Comment by HorlogeSkynet** [comment]
I started doing something like, let me try again then.

--> It won't be backward compatible if we change the `message` field though

**Comment by ADmad** [comment]
It can be done in a BC way. Just check if message it set and if so overwrite response.body.

**Comment by HorlogeSkynet** [comment]
I gotta go, could you tell me whether it is the idea you got ?
If so, I'll try to finish this later.

File: src/Routing/Filter/ThrottleFilter.php [hunk 1]
```diff
@@ -46,10 +46,16 @@ public function beforeDispatch(Event $event)
 
         // client has reached rate limit
         $event->stopPropagation();
-        $response = new Response(['body' => $this->config('message')]);
+        $response = new Response(['body' => (null !== $this->config('message') ? $this->config('message') : $this->config('response')['body'])]);
```
Comment (ADmad) [inline src/Routing/Filter/ThrottleFilter.php]: This line is not very readable. Better to assign a `$message` variable first and then use it. Also because don't use yoda comparisons.

File: tests/TestCase/Middleware/ThrottleMiddlewareTest.php [hunk 2]
```diff
@@ -38,6 +38,7 @@ public function testConstructor()
         $this->assertEquals('Rate limit exceeded', $result['message']);
         $this->assertEquals('+1 minute', $result['interval']);
         $this->assertEquals(10, $result['limit']);
+        $this->assertEquals('text/html', $result['type']);
```
Comment (ADmad) [inline tests/TestCase/Middleware/ThrottleMiddlewareTest.php]: You should also add a new test with custom headers set.

**Review by ADmad — CHANGES_REQUESTED** [review]

**Comment by HorlogeSkynet** [comment]
@ADmad sorry for the multiple notifications, 'got a problem with `phpunit` and `APCU`, coulndn't test on my system...
Tell me what you think about latest changes, I've set back the `type` handling because Cake back-end overwrites the headers while setting the `contentType`...

EDIT : Don't know why a header is missing with PHP5...

**Comment by ADmad** [comment]
I have re-targeted your PR to new `dev` branch which will be CakePHP 3.4+. It has some cleanup and the test suite fixes to run tests on travis. You'll need to fix the merge conflicts.

File: src/Routing/Filter/ThrottleFilter.php [hunk 3]
```diff
@@ -35,20 +35,35 @@ public function beforeDispatch(Event $event)
         $this->_initCache();
         $this->_count = $this->_touch();
 
+        $config = $this->getConfig();
+
         // client has not exceeded rate limit
-        if ($this->_count <= $this->getConfig('limit')) {
-            $this->_setHeaders($event->getData('response'));
+        if ($this->_count <= $config['limit']) {
+            $this->_setHeaders($event->getData['response']);
```
Comment (ADmad) [inline src/Routing/Filter/ThrottleFilter.php]: `getData()` is a method not array :)

Comment (HorlogeSkynet) [reply src/Routing/Filter/ThrottleFilter.php]: :rofl: This why we don't have to use GitHub, sorry :joy:

File: src/Middleware/ThrottleMiddleware.php [hunk 4]
```diff
@@ -45,11 +45,25 @@ public function __invoke(ServerRequestInterface $request, ResponseInterface $res
...
+            }
+
+            if (isset($config['message'])) {
+                $stream->write((string)$config['message']);
+            } else {
+                $stream->write((string)$config['response']['body']);
+            }
 
             return $response->withStatus(429)
+                ->withType($config['response']['type'])
```
Comment (ADmad) [inline src/Middleware/ThrottleMiddleware.php]: `withType()` is not a method of `ResponseInterface` so it would be better to avoid using it. Also using this method makes [phpstan](https://travis-ci.org/UseMuffin/Throttle/jobs/359511902#L603) unhappy as the `$response` argument is typehinted as `Psr\Http\Message\ResponseInterface` not `Cake\Http\Response`. 

Using a separate `type` key inside `response` isn't really necessary anyway. One can just set `Content-Type` in `headers`.

File: src/Routing/Filter/ThrottleFilter.php [hunk 5]
```diff
@@ -35,20 +35,35 @@ public function beforeDispatch(Event $event)
         $this->_initCache();
         $this->_count = $this->_touch();
 
+        $config = $this->getConfig();
+
         // client has not exceeded rate limit
-        if ($this->_count <= $this->getConfig('limit')) {
-            $this->_setHeaders($event->getData('response'));
+        if ($this->_count <= $config['limit']) {
+            $this->_setHeaders($event->getData()['response']);
```
Comment (ADmad) [inline src/Routing/Filter/ThrottleFilter.php]: Keep the line `$this->_setHeaders($event->getData('response'));` as it was earlier.

File: src/Middleware/ThrottleMiddleware.php (hunk 4 above)
Comment (HorlogeSkynet) [reply src/Middleware/ThrottleMiddleware.php]: Well as I've explained sooner today, when we set the stream with the `withBody` method, Cake overwrites `Content-Type` header... That's why I've set back the `type` handling

Comment (ADmad) [reply src/Middleware/ThrottleMiddleware.php]: Okay, then let's just remove the typehints for arguments of `__invoke()` method. Since the method isn't going to be called directly by user the typehints aren't really useful anyway. Also update the docs for request and response argument to use `Cake\Http\ServerRequest|Reponse` instead of the interfaces.

P. S. Don't forget to remove the "use" statements for the interfaces at top of file :)

Comment (HorlogeSkynet) [reply src/Middleware/ThrottleMiddleware.php]: I may have a solution, lemme 5 minutes

--> Well let's go then :rofl:

Comment (ADmad) [reply src/Middleware/ThrottleMiddleware.php]: Yeah this middleware isn't really framework agnostic so trying to stick to the PSR7 interfaces doesn't help.

Comment (HorlogeSkynet) [reply src/Middleware/ThrottleMiddleware.php]: I've tried something (sorry for the whole process, first time with PSR7 stuff :grimacing:)

**Comment by ADmad** [comment]
Cake 3.4 doesn't have `Response::getType()` :stuck_out_tongue: 
https://travis-ci.org/UseMuffin/Throttle/jobs/359526632#L551

Let's just bump to cake 3.5. Then we can also use `Response::withStringBody()` and avoid having to create `Stream` instance.

**Comment by HorlogeSkynet** [comment]
I totally agree ! It's endless :rofl:

File: src/Middleware/ThrottleMiddleware.php [hunk 6]
```diff
@@ -34,23 +34,35 @@ public function __construct($config = [])
...
      */
-    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, callable $next)
+    public function __invoke(ServerRequest $request, Response $response, callable $next)
     {
         $this->_setIdentifier($request);
         $this->_initCache();
         $this->_count = $this->_touch();
 
-        if ($this->_count > $this->getConfig('limit')) {
-            $stream = new Stream('php://memory', 'wb+');
```
Comment (ADmad) [inline src/Middleware/ThrottleMiddleware.php:49]: You need to remove the unneeded "use" statement for `Stream` too.

**Comment by ADmad** [comment]
You need to now update phpunit constraint to `"^5.7|^6.0"` to get this build to pass https://travis-ci.org/UseMuffin/Throttle/jobs/359529770.

**Comment by ADmad** [comment]
I hope this will be last change required. Else I am going to takeover and finish this myself :laughing:.

**Comment by HorlogeSkynet** [comment]
Sorry man but first time with PHPUNIT too... Didn't know there were version conflicts about it, and the whole thing went well on my system :confused: 

Oups sorry, didn't see the checkbox for maintainers edition was unchecked :thinking: :thinking:

**Comment by ADmad** [comment]
It's not your fault :). A great deal of versioning dance is required to maintain compatibility between various versions of php, cake and phpunit.

**Comment by ADmad** [comment]
I'll fix that phpunit issue. Thanks for seeing this through. I'll make a new stable release off the `dev` branch once Cake 3.6 stable is released (which should be soon), so as to ensure not to hit any new deprecations.

**Comment by HorlogeSkynet** [comment]
Thanks you for your help and your guidance.
I'll check that :ok_hand: 

See you soon.

**Comment by ADmad** [comment]
All green now https://travis-ci.org/UseMuffin/Throttle/builds/359539066.

**Comment by bravo-kernel** [comment]
Good job @HorlogeSkynet and.... mr. @ADmad, you are truly a gentleman 👏
